[OpenMP] Remove optimization skipping reduction struct initialization

rodrigo-ceccato opened the pull request:
This commit removes an optimization that skips the initialization of the
reduction struct if the number of threads in a team is 1. This optimization
caused a bug with Hidden Helper Threads. When the task group is initially
initialized by the master thread but a Hidden Helper Thread executes a target
nowait region, it requires the reduction struct initialization to properly
accumulate the data.

This commit also adds a LIT test for issue #57522 to ensure that the issue is
properly addressed and that the optimization removal does not introduce any
regressions.

Fixes: #57522

rodrigo-ceccato committed:
[OpenMP] Ensure initialization of reduction struct in taskloop

This commit ensures the initialization of the reduction struct in task loops,
even when the number of threads in a team is 1, provided that hidden helper
threads are active. Previously, an optimization skipped this initialization
in cases where there was only a single thread in the team, which led to a bug.

The bug behavior resulted in a segmentation fault when hidden helper threads
were active. This occurred because the reduction struct was not initialized
when the task group was created by a team with only one thread, causing the
hidden helper thread to encounter issues when accessing the reduction data.

In addition to this fix, a LIT test for issue #57522 is included to validate
the resolution and prevent any regressions.

Fixes: #57522

shiltian commented on openmp/runtime/src/kmp_tasking.cpp:
```diff
@@ -2699,6 +2694,7 @@ void *__kmpc_task_reduction_get_th_data(int gtid, void *tskgrp, void *data) {
         return p_priv[tid];
       }
     }
+    KMP_ASSERT(tg->parent != NULL);
```
```suggestion
    KMP_ASSERT(tg->parent);
```

rodrigo-ceccato replied:
Done.

shiltian commented on openmp/runtime/src/kmp_tasking.cpp:
```diff
@@ -2512,11 +2512,6 @@ void *__kmp_task_reduction_init(int gtid, int num, T *data) {
   KMP_ASSERT(tg != NULL);
   KMP_ASSERT(data != NULL);
   KMP_ASSERT(num > 0);
-  if (nth == 1) {
```
Instead of completely removing the branch, it might be better to check if HHT is enabled. If no, then we can safely return here.

jdoerfert replied:
I'm not sure how important single threaded reduction costs really are, but sure.

rodrigo-ceccato commented:
Squashed and updated the commit message.


```
ninja check-openmp
[0/1] Running OpenMP tests
-- Testing: 724 tests, 144 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 

Testing Time: 144.14s
  Unsupported: 100
  Passed     : 624
  ```
